### PR TITLE
improve task migration skill

### PR DESCRIPTION
**Pull request by JanProvaznik**
Merge adversarial compatibility review guidance into the existing migration skill:
- 6 Deadly Compatibility Sins (real bugs found during task migrations)
- Red-team audit protocol with edge-case analysis tables
- Compatibility test generation matrix
- Unified sign-off checklist covering migration + compatibility

### CODE REVIEW

File: .github/skills/multithreaded-task-migration/SKILL.md [hunk 1]
```diff
@@ -27,164 +23,73 @@ public class MyTask : Task, IMultiThreadableTask
...
-`GetAbsolutePath()` throws for null/empty inputs. See [Exception Handling in Batch Operations](#exception-handling-in-batch-operations) for handling strategies.
-
 The [`AbsolutePath`](https://github.com/dotnet/msbuild/blob/main/src/Framework/PathHelpers/AbsolutePath.cs) struct:
-- Has `Value` property returning the absolute path string
-- Has `OriginalValue` property preserving the input path  
-- Is implicitly convertible to `string` for File/Directory API compatibility
+- `Value` — the absolute path string
+- `OriginalValue` — preserves the input path (use for error messages and `[Output]` properties)
+- Implicitly convertible to `string` for File/Directory API compatibility
+- `GetCanonicalForm()` — resolves `..` segments and normalizes separators (see [Sin 5](#sin-5-canonicalization-mismatch))
```
Comment (rainersigwald) [inline .github/skills/multithreaded-task-migration/SKILL.md:44]: Why is it `GetCanonicalForm()` instead of our more usual `Normalize()`?

**Review by rainersigwald — COMMENTED** [review]
LGTM if it helps--my gut instinct is it's a bit wordy for context windows but I haven't really updated for the new models.

File: .github/skills/multithreaded-task-migration/SKILL.md [hunk 2]
```diff
@@ -1,22 +1,18 @@
...
-MSBuild's multithreaded execution model requires tasks to avoid global process state (working directory, environment variables). Thread-safe tasks declare this capability by annotating with  `MSBuildMultiThreadableTask` and use `TaskEnvironment` provided by `IMultiThreadableTask` for safe alternatives.
+MSBuild's multithreaded execution model requires tasks to avoid global process state (working directory, environment variables). Thread-safe tasks declare this capability via `MSBuildMultiThreadableTask` and use `TaskEnvironment` from `IMultiThreadableTask` for safe alternatives.
 
 ## Migration Steps
 
 ### Step 1: Update Task Class Declaration
 
-a. add the attribute
-b. AND implement the interface if it's necessary to use TaskEnvironment APIs.
+a. Add the attribute.
```
Comment (JanKrivanek) [inline .github/skills/multithreaded-task-migration/SKILL.md]:
```suggestion
a. Ensure the task implementing class is decorated with the `MSBuildMultiThreadableTask` attribute.
```

I know the example is below - but being specific doesn't hurt

File: .github/skills/multithreaded-task-migration/SKILL.md [hunk 3]
```diff
@@ -27,164 +23,73 @@ public class MyTask : Task, IMultiThreadableTask
...
 ```csharp
-// BEFORE - File.Exists uses process working directory for relative paths (UNSAFE)
-if (File.Exists(inputPath))
-{
-    string content = File.ReadAllText(inputPath);
-}
-
-// AFTER - Absolutize first, then use in file operations (SAFE)
 AbsolutePath absolutePath = TaskEnvironment.GetAbsolutePath(inputPath);
 if (File.Exists(absolutePath))
```
Comment (JanKrivanek) [inline .github/skills/multithreaded-task-migration/SKILL.md:34]: Unrelated to this PR - have you considered banning System.IO APIs and providing wrappers accepting only the strong typed paths?

It is extra burden for the conversion - but it should be straightworward and would add extra safety + aid testing

File: .github/skills/multithreaded-task-migration/SKILL.md [hunk 4]
```diff
@@ -27,164 +23,73 @@ public class MyTask : Task, IMultiThreadableTask
...
-// BEFORE (UNSAFE)
-string value = Environment.GetEnvironmentVariable("VAR");
-Environment.SetEnvironmentVariable("VAR", "value");
-
-// AFTER (SAFE)
-string value = TaskEnvironment.GetEnvironmentVariable("VAR");
-TaskEnvironment.SetEnvironmentVariable("VAR", "value");
+// BEFORE (UNSAFE)                              // AFTER (SAFE)
+Environment.GetEnvironmentVariable("VAR");      TaskEnvironment.GetEnvironmentVariable("VAR");
+Environment.SetEnvironmentVariable("VAR", "v"); TaskEnvironment.SetEnvironmentVariable("VAR", "v");
```
Comment (JanKrivanek) [inline .github/skills/multithreaded-task-migration/SKILL.md]:
```suggestion
| BEFORE (UNSAFE)                                  | AFTER (SAFE)                                       |
|--------------------------------------------------|----------------------------------------------------|
| Environment.GetEnvironmentVariable("VAR");       | TaskEnvironment.GetEnvironmentVariable("VAR");     |
| Environment.SetEnvironmentVariable("VAR", "v");  | TaskEnvironment.SetEnvironmentVariable("VAR", "v"); |
```

Consider markup format for better clarity

Comment (JanKrivanek) [reply .github/skills/multithreaded-task-migration/SKILL.md]: Same in further examples

File: .github/skills/multithreaded-task-migration/SKILL.md (hunk 3 above)
Comment (JanKrivanek) [reply .github/skills/multithreaded-task-migration/SKILL.md:34]: On the other hand it might provide false feeling of safety, as it doesn't prevent flowing basic `strings` through chain of calls ending in IO in unrelated assemblies.  - so just disregard :)

File: .github/skills/multithreaded-task-migration/SKILL.md [hunk 5]
```diff
@@ -27,164 +23,73 @@ public class MyTask : Task, IMultiThreadableTask
...
+| **Use TaskEnvironment** | `Environment.CurrentDirectory`, `Get/SetEnvironmentVariable`, `Path.GetFullPath`, `ProcessStartInfo` | See Steps 2-4 |
+| **Need absolute paths** | `File.*`, `Directory.*`, `FileInfo`, `DirectoryInfo`, `FileStream`, `StreamReader/Writer` | Absolutize first |
+| **Review required** | `Assembly.Load*`, `Activator.CreateInstance*` | Check for version conflicts |
 
 ## Practical Notes
 
 ### CRITICAL: Trace All Path String Usage
 
-**You MUST trace every path string variable through the entire codebase** to find all places where it flows into file system operations - including helper methods, utility classes, and third-party code that may internally use File APIs.
+Trace every path string through all method calls and assignments to find all places it flows into file system operations — including helper methods that may internally use File APIs.
```
Comment (JanKrivanek) [inline .github/skills/multithreaded-task-migration/SKILL.md]: "File APIs." --> "File System APIs." ?

File: .github/skills/multithreaded-task-migration/SKILL.md [hunk 6]
```diff
@@ -197,48 +102,192 @@ foreach (ITaskItem item in SourceFiles)
...
+
+```csharp
+// CORRECT: hoist above try so catch can use it too
+AbsolutePath abs = TaskEnvironment.GetAbsolutePath(OutputManifest.ItemSpec);
+try {
+    WriteFile(abs);
+} catch (Exception ex) {
+    string lockMsg = LockCheck.GetLockedFileMessage(abs);        // absolute → correct file
+    Log.LogError("Failed: {0}", OutputManifest.ItemSpec, ...);   // original → user-friendly
+}
```
Comment (JanKrivanek) [inline .github/skills/multithreaded-task-migration/SKILL.md:195]: Prefer giving it example formatted in more likely expected way - to minimize need for model to roundtrip through compiler warnings on styling.

File: .github/skills/multithreaded-task-migration/SKILL.md (hunk 1 above)
Comment (JanProvaznik) [reply .github/skills/multithreaded-task-migration/SKILL.md:44]: we discussed this with Chet and he was not opposed to that naming
